File: xcode/tests/tokenTest.m
/*
 *	Generated 05/26/2015 13:40:16
 */


/*
 This is free and unencumbered software released into the public domain.
 
 Anyone is free to copy, modify, publish, use, compile, sell, or
 distribute this software, either in source code form or as a compiled
 binary, for any purpose, commercial or non-commercial, and by any
 means.
 
 In jurisdictions that recognize copyright laws, the author or authors
 of this software dedicate any and all copyright interest in the
 software to the public domain. We make this dedication for the benefit
 of the public at large and to the detriment of our heirs and
 successors. We intend this dedication to be an overt act of
 relinquishment in perpetuity of all present and future rights to this
 software under copyright law.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 OTHER DEALINGS IN THE SOFTWARE.
 
 For more information, please refer to <http://unlicense.org/>
 */


#import <XCTest/XCTest.h>
#import <libCmn/libCmn.h>


// All code under test must be linked into the Unit Test bundle
// Test Macros:
//      STAssert(expression, failure_description, ...)
//      STFail(failure_description, ...)
//      STAssertEqualObjects(object_1, object_2, failure_description, ...) uses isEqualTo:
//      STAssertEquals(value_1, value_2, failure_description, ...)
//      STAssertEqualsWithAccuracy(value_1, value_2, accuracy, failure_description, ...)
//      STAssertNil(expression, failure_description, ...)
//      STAssertNotNil(expression, failure_description, ...)
//      STAssertTrue(expression, failure_description, ...)
//      STAssertFalse(expression, failure_description, ...)
//      STAssertThrows(expression, failure_description, ...)
//      STAssertThrowsSpecific(expression, exception_class, failure_description, ...)
//      STAssertThrowsSpecificNamed(expression, exception_class, exception_name, failure_description, ...)
//      STAssertNoThrow(expression, failure_description, ...)
//      STAssertNoThrowSpecific(expression, exception_class, failure_description, ...)
//      STAssertNoThrowSpecificNamed(expression, exception_class, exception_name, failure_description, ...)


#include    "token_internal.h"
#include    "szTbl.h"



@interface tokenTests : XCTestCase

@end

@implementation tokenTests


- (void)setUp
{
    [super setUp];
    // Put setup code here. This method is called before the invocation of each
    // test method in the class.
    
    mem_Init( );
    
}


- (void)tearDown
{
    mem_Dump( );
    mem_Release( );

    // Put teardown code here. This method is called after the invocation of each
    // test method in the class.
    [super tearDown];
}




- (void)testOpenClose
{
    TOKEN_DATA	*pObj = OBJ_NIL;
   
    pObj = token_Alloc();
    XCTAssertFalse( (OBJ_NIL == pObj) );
    pObj = token_Init( pObj );
    XCTAssertFalse( (OBJ_NIL == pObj) );
    if (pObj) {
        obj_Release(pObj);
        pObj = OBJ_NIL;
    }

    szTbl_SharedReset();
}



- (void)testOpenClose01
{
    TOKEN_DATA	*pObj = OBJ_NIL;
    ASTR_DATA   *pStr;
    const
    char        *pJSON_Con = "{\"objectType\":\"token\","
                                "\"fileName\":\"abc\","
                                "\"lineNo\":52,"
                                "\"colNo\":22,"
                                "\"cls\":11,"
                                "\"type\":\"NUMBER\","
                                "\"data\":64}\n";
    ERESULT     eRc;
    TOKEN_DATA  *pToken;
    
    pObj = token_Alloc();
    XCTAssertFalse( (OBJ_NIL == pObj) );
    pObj = token_InitInteger(pObj, "abc", 52, 22, 11, 64);
    XCTAssertFalse( (OBJ_NIL == pObj) );
    if (pObj) {
        
        XCTAssertTrue( (0 == strcmp("abc", token_getFileName(pObj))) );
        XCTAssertTrue( (52 == token_getLineNo(pObj)) );
        XCTAssertTrue( (22 == token_getColNo(pObj)) );
        XCTAssertTrue( (11 == token_getClass(pObj)) );
        XCTAssertTrue( (64 == token_getInteger(pObj)) );
        
        pStr = token_ToJSON(pObj);
        XCTAssertFalse( (OBJ_NIL == pStr) );
        //fprintf(stderr, "%s\n", AStr_getData(pStr));
        eRc = AStr_CompareA(pStr, pJSON_Con);
        XCTAssertTrue( (ERESULT_SUCCESS_EQUAL == eRc) );
        
        pToken = token_NewFromJSONString(pStr);
        XCTAssertFalse( (OBJ_NIL == pToken) );
        XCTAssertTrue( (0 == strcmp("abc", token_getFileName(pToken))) );
        XCTAssertTrue( (52 == token_getLineNo(pToken)) );
        XCTAssertTrue( (22 == token_getColNo(pToken)) );
        XCTAssertTrue( (11 == token_getClass(pToken)) );
        XCTAssertTrue( (64 == token_getInteger(pToken)) );
        obj_Release(pStr);
        pStr = OBJ_NIL;

        pStr = token_ToJSON(pToken);
        XCTAssertFalse( (OBJ_NIL == pStr) );
        //fprintf(stderr, "%s\n", AStr_getData(pStr));
        eRc = AStr_CompareA(pStr, pJSON_Con);
        XCTAssertTrue( (ERESULT_SUCCESS_EQUAL == eRc) );
        
        obj_Release(pToken);
        pToken = OBJ_NIL;
        obj_Release(pStr);
        pStr = OBJ_NIL;
        obj_Release(pObj);
        pObj = OBJ_NIL;
    }
    
    szTbl_SharedReset();
}



- (void)testOpenClose02
{
    TOKEN_DATA	*pObj = OBJ_NIL;
    ASTR_DATA   *pStr;
    const
    char        *pJSON_Con = "{\"objectType\":\"token\","
    "\"fileName\":\"abc\","
    "\"lineNo\":52,"
    "\"colNo\":22,"
    "\"cls\":11,"
    "\"type\":\"CHAR\","
    "\"data\":90}\n";
    ERESULT     eRc;
    TOKEN_DATA  *pToken;
    
    pObj = token_Alloc();
    XCTAssertFalse( (OBJ_NIL == pObj), @"Could not alloc  TOKEN_DATA" );
    pObj = token_InitCharW(pObj, "abc", 52, 22, 11, 'Z');
    XCTAssertFalse( (OBJ_NIL == pObj), @"" );
    if (pObj) {
        
        XCTAssertTrue( (0 == strcmp("abc", token_getFileName(pObj))), @"" );
        XCTAssertTrue( (52 == token_getLineNo(pObj)), @"" );
        XCTAssertTrue( (22 == token_getColNo(pObj)), @"" );
        XCTAssertTrue( (11 == token_getClass(pObj)), @"" );
        XCTAssertTrue( ('Z' == token_getChrW(pObj)), @"" );
        
        pStr = token_ToJSON(pObj);
        XCTAssertFalse( (OBJ_NIL == pStr), @"" );
        fprintf(stderr, "%s\n", AStr_getData(pStr));
        eRc = AStr_CompareA(pStr, pJSON_Con);
        XCTAssertTrue( (ERESULT_SUCCESS_EQUAL == eRc), @"" );
        
        pToken = token_NewFromJSONString(pStr);
        XCTAssertFalse( (OBJ_NIL == pToken), @"" );
        XCTAssertTrue( (0 == strcmp("abc", token_getFileName(pToken))), @"" );
        XCTAssertTrue( (52 == token_getLineNo(pToken)), @"" );
        XCTAssertTrue( (22 == token_getColNo(pToken)), @"" );
        XCTAssertTrue( (11 == token_getClass(pToken)), @"" );
        XCTAssertTrue( ('Z' == token_getChrW(pObj)), @"" );
        obj_Release(pStr);
        pStr = OBJ_NIL;
        
        pStr = token_ToJSON(pToken);
        XCTAssertFalse( (OBJ_NIL == pStr), @"" );
        //fprintf(stderr, "%s\n", AStr_getData(pStr));
        eRc = AStr_CompareA(pStr, pJSON_Con);
        XCTAssertTrue( (ERESULT_SUCCESS_EQUAL == eRc), @"" );
        
        obj_Release(pToken);
        pToken = OBJ_NIL;
        obj_Release(pStr);
        pStr = OBJ_NIL;
        obj_Release(pObj);
        pObj = OBJ_NIL;
    }
    
    szTbl_SharedReset();
}



- (void)testOpenClose03
{
    TOKEN_DATA	*pObj = OBJ_NIL;
    ASTR_DATA   *pStr;
    const
    char        *pJSON_Con = "{\"objectType\":\"token\","
    "\"fileName\":\"abc\","
    "\"lineNo\":52,"
    "\"colNo\":22,"
    "\"cls\":11,"
    "\"type\":\"STRING\","
    "\"data\":\"12345\"}\n";
    ERESULT     eRc;
    TOKEN_DATA  *pToken;
    WSTR_DATA   *pStrW;
    
    pObj = token_Alloc();
    XCTAssertFalse( (OBJ_NIL == pObj) );
    pObj = token_InitStrW(pObj, "abc", 52, 22, 11, L"12345");
    XCTAssertFalse( (OBJ_NIL == pObj) );
    if (pObj) {
        
        XCTAssertTrue( (0 == strcmp("abc", token_getFileName(pObj))) );
        XCTAssertTrue( (52 == token_getLineNo(pObj)) );
        XCTAssertTrue( (22 == token_getColNo(pObj)) );
        XCTAssertTrue( (11 == token_getClass(pObj)) );
        pStrW = token_getStringW(pObj);
        XCTAssertFalse( (OBJ_NIL == pStrW) );
        XCTAssertTrue( (OBJ_IDENT_WSTR == obj_getType(pStrW)) );
        eRc = WStr_CompareW(pStrW,L"12345");
        XCTAssertTrue( (ERESULT_SUCCESS_EQUAL == eRc), @"" );
        
        pStr = token_ToJSON(pObj);
        XCTAssertFalse( (OBJ_NIL == pStr), @"" );
        fprintf(stderr, "%s\n", AStr_getData(pStr));
        eRc = AStr_CompareA(pStr, pJSON_Con);
        XCTAssertTrue( (ERESULT_SUCCESS_EQUAL == eRc), @"" );
        
        pToken = token_NewFromJSONString(pStr);
        XCTAssertFalse( (OBJ_NIL == pToken), @"" );
        XCTAssertTrue( (0 == strcmp("abc", token_getFileName(pToken))), @"" );
        XCTAssertTrue( (52 == token_getLineNo(pToken)), @"" );
        XCTAssertTrue( (22 == token_getColNo(pToken)), @"" );
        XCTAssertTrue( (11 == token_getClass(pToken)), @"" );
        eRc = WStr_CompareW(token_getStringW(pObj),L"12345");
        XCTAssertTrue( (ERESULT_SUCCESS_EQUAL == eRc), @"" );
        obj_Release(pStr);
        pStr = OBJ_NIL;
        
        pStr = token_ToJSON(pToken);
        XCTAssertFalse( (OBJ_NIL == pStr), @"" );
        //fprintf(stderr, "%s\n", AStr_getData(pStr));
        eRc = AStr_CompareA(pStr, pJSON_Con);
        XCTAssertTrue( (ERESULT_SUCCESS_EQUAL == eRc), @"" );
        
        obj_Release(pToken);
        pToken = OBJ_NIL;
        obj_Release(pStr);
        pStr = OBJ_NIL;
        obj_Release(pObj);
        pObj = OBJ_NIL;
    }
    
    szTbl_SharedReset();
}



- (void)testOpenClose04
{
    TOKEN_DATA	*pObj = OBJ_NIL;
    ASTR_DATA   *pStr;
    const
    char        *pJSON_Con = "{\"objectType\":\"token\","
    "\"fileName\":\"abc\","
    "\"lineNo\":52,"
    "\"colNo\":22,"
    "\"cls\":11,"
    "\"type\":\"STRING\","
    "\"data\":\"12345\"}\n";
    ERESULT     eRc;
    TOKEN_DATA  *pToken;
    WSTR_DATA   *pStrW;
    
    pObj = token_Alloc();
    XCTAssertFalse( (OBJ_NIL == pObj) );
    pObj = token_InitStrA(pObj, "abc", 52, 22, 11, "12345");
    XCTAssertFalse( (OBJ_NIL == pObj) );
    if (pObj) {
        
        XCTAssertTrue( (0 == strcmp("abc", token_getFileName(pObj))) );
        XCTAssertTrue( (52 == token_getLineNo(pObj)) );
        XCTAssertTrue( (22 == token_getColNo(pObj)) );
        XCTAssertTrue( (11 == token_getClass(pObj)) );
        pStrW = token_getStringW(pObj);
        XCTAssertFalse( (OBJ_NIL == pStrW) );
        XCTAssertTrue( (OBJ_IDENT_WSTR == obj_getType(pStrW)) );
        eRc = WStr_CompareW(pStrW,L"12345");
        XCTAssertTrue( (ERESULT_SUCCESS_EQUAL == eRc) );
        
        pStr = token_ToJSON(pObj);
        XCTAssertFalse( (OBJ_NIL == pStr) );
        fprintf(stderr, "%s\n", AStr_getData(pStr));
        eRc = AStr_CompareA(pStr, pJSON_Con);
        XCTAssertTrue( (ERESULT_SUCCESS_EQUAL == eRc) );
        
        pToken = token_NewFromJSONString(pStr);
        XCTAssertFalse( (OBJ_NIL == pToken) );
        XCTAssertTrue( (0 == strcmp("abc", token_getFileName(pToken))) );
        XCTAssertTrue( (52 == token_getLineNo(pToken)) );
        XCTAssertTrue( (22 == token_getColNo(pToken)) );
        XCTAssertTrue( (11 == token_getClass(pToken)) );
        eRc = WStr_CompareW(token_getStringW(pObj),L"12345");
        XCTAssertTrue( (ERESULT_SUCCESS_EQUAL == eRc) );
        obj_Release(pStr);
        pStr = OBJ_NIL;
        
        pStr = token_ToJSON(pToken);
        XCTAssertFalse( (OBJ_NIL == pStr) );
        //fprintf(stderr, "%s\n", AStr_getData(pStr));
        eRc = AStr_CompareA(pStr, pJSON_Con);
        XCTAssertTrue( (ERESULT_SUCCESS_EQUAL == eRc) );
        
        obj_Release(pToken);
        pToken = OBJ_NIL;
        obj_Release(pStr);
        pStr = OBJ_NIL;
        obj_Release(pObj);
        pObj = OBJ_NIL;
    }
    
    szTbl_SharedReset();
}



@end

File: tests/Token_test.c
// vi:nu:et:sts=4 ts=4 sw=4
/*
 *	Generated 01/02/2020 15:58:09
 */





// All code under test must be linked into the Unit Test bundle
// Test Macros:
//      TINYTEST_ASSERT(condition)
//      TINYTEST_ASSERT_MSG(condition,msg)
//      TINYTEST_EQUAL(expected, actual)
//      TINYTEST_EQUAL_MSG(expected, actual, msg)
//      TINYTEST_FALSE_MSG(condition,msg)
//      TINYTEST_FALSE(condition)
//      TINYTEST_TRUE_MSG(pointer,msg)
//      TINYTEST_TRUE(condition)





#include    <tinytest.h>
#include    <cmn_defs.h>
#include    <SrcErrors.h>
#include    <szTbl.h>
#include    <trace.h>
#include    <Token_internal.h>



int             setUp(
    const
    char            *pTestName
)
{
    mem_Init( );
    trace_Shared( ); 
    // Put setup code here. This method is called before the invocation of each
    // test method in the class.
    
    return 1; 
}


int             tearDown(
    const
    char            *pTestName
)
{
    // Put teardown code here. This method is called after the invocation of each
    // test method in the class.

    SrcErrors_SharedReset();
    szTbl_SharedReset();
    trace_SharedReset( );
    if (mem_Dump( ) ) {
        fprintf(
                stderr,
                "\x1b[1m"
                "\x1b[31m"
                "ERROR: "
                "\x1b[0m"
                "Leaked memory areas were found!\n"
        );
        exitCode = 4;
        return 0;
    }
    mem_Release( );
    
    return 1; 
}






int             test_Token_OpenClose(
    const
    char            *pTestName
)
{
    ERESULT         eRc = ERESULT_SUCCESS;
    TOKEN_DATA	    *pObj = OBJ_NIL;
    bool            fRc;
   
    fprintf(stderr, "Performing: %s\n", pTestName);

    pObj = Token_Alloc( );
    TINYTEST_FALSE( (OBJ_NIL == pObj) );
    pObj = Token_Init( pObj );
    TINYTEST_FALSE( (OBJ_NIL == pObj) );
    if (pObj) {

        //obj_TraceSet(pObj, true);       
        fRc = obj_IsKindOf(pObj, OBJ_IDENT_TOKEN);
        TINYTEST_TRUE( (fRc) );
        
        // Test something.
        TINYTEST_FALSE( (ERESULT_FAILED(eRc)) );

        obj_Release(pObj);
        pObj = OBJ_NIL;
    }

    fprintf(stderr, "...%s completed.\n\n\n", pTestName);
    return 1;
}



int             test_Token_Copy01 (
    const
    char            *pTestName
)
{
    ERESULT         eRc = ERESULT_SUCCESS;
    TOKEN_DATA       *pObj1 = OBJ_NIL;
    TOKEN_DATA       *pObj2 = OBJ_NIL;
    bool            fRc;
#if defined(TOKEN_JSON_SUPPORT)
    ASTR_DATA       *pStr = OBJ_NIL;
#endif
    SRCLOC          loc = {1,2,0,3,4};

    fprintf(stderr, "Performing: %s\n", pTestName);

    pObj1 = Token_NewInteger(&loc, 11, 64);
    TINYTEST_FALSE( (OBJ_NIL == pObj1) );
    if (pObj1) {

        //obj_TraceSet(pObj1, true);
        fRc = obj_IsKindOf(pObj1, OBJ_IDENT_TOKEN);
        TINYTEST_TRUE( (fRc) );
        XCTAssertTrue( ( 1 == Token_getFileIndex(pObj1)) );
        XCTAssertTrue( ( 4 == Token_getOffset(pObj1)) );
        XCTAssertTrue( ( 3 == Token_getLineNo(pObj1)) );
        XCTAssertTrue( ( 2 == Token_getColNo(pObj1)) );
        XCTAssertTrue( (11 == Token_getClass(pObj1)) );
        XCTAssertTrue( (TOKEN_TYPE_INTEGER == Token_getType(pObj1)) );
        XCTAssertTrue( (64 == Token_getInteger(pObj1)) );
        {
            ASTR_DATA       *pStr = Token_ToDebugString(pObj1, 4);
            if (pStr) {
                fprintf(stderr, "Debug: %s\n", AStr_getData(pStr));
                obj_Release(pStr);
                pStr = OBJ_NIL;
            }
        }


        // Test assign.
        pObj2 = Token_New();
        TINYTEST_FALSE( (OBJ_NIL == pObj2) );
        eRc = Token_Assign(pObj1, pObj2);
        TINYTEST_FALSE( (ERESULT_FAILED(eRc)) );

        fRc = obj_IsKindOf(pObj2, OBJ_IDENT_TOKEN);
        TINYTEST_TRUE( (fRc) );
        XCTAssertTrue( ( 1 == Token_getFileIndex(pObj2)) );
        XCTAssertTrue( ( 4 == Token_getOffset(pObj2)) );
        XCTAssertTrue( ( 3 == Token_getLineNo(pObj2)) );
        XCTAssertTrue( ( 2 == Token_getColNo(pObj2)) );
        XCTAssertTrue( (11 == Token_getClass(pObj2)) );
        XCTAssertTrue( (TOKEN_TYPE_INTEGER == Token_getType(pObj2)) );
        XCTAssertTrue( (64 == Token_getInteger(pObj2)) );

        obj_Release(pObj2);
        pObj2 = OBJ_NIL;

        // Test copy.
        pObj2 = Token_Copy(pObj1);
        TINYTEST_FALSE( (OBJ_NIL == pObj2) );

        fRc = obj_IsKindOf(pObj2, OBJ_IDENT_TOKEN);
        TINYTEST_TRUE( (fRc) );
        XCTAssertTrue( ( 1 == Token_getFileIndex(pObj2)) );
        XCTAssertTrue( ( 4 == Token_getOffset(pObj2)) );
        XCTAssertTrue( ( 3 == Token_getLineNo(pObj2)) );
        XCTAssertTrue( ( 2 == Token_getColNo(pObj2)) );
        XCTAssertTrue( (11 == Token_getClass(pObj2)) );
        XCTAssertTrue( (TOKEN_TYPE_INTEGER == Token_getType(pObj2)) );
        XCTAssertTrue( (64 == Token_getInteger(pObj2)) );

        obj_Release(pObj2);
        pObj2 = OBJ_NIL;

        // Test json support.
#if defined(TOKEN_JSON_SUPPORT)
        pStr = Token_ToJson(pObj1);
        TINYTEST_FALSE( (OBJ_NIL == pStr) );
        fprintf(stderr, "JSON: %s\n", AStr_getData(pStr));
        pObj2 = Token_NewFromJsonString(pStr);
        TINYTEST_FALSE( (OBJ_NIL == pObj2) );
        fRc = obj_IsKindOf(pObj2, OBJ_IDENT_TOKEN);
        TINYTEST_TRUE( (fRc) );
        obj_Release(pStr);
        pStr = OBJ_NIL;
        XCTAssertTrue( ( 1 == Token_getFileIndex(pObj2)) );
        XCTAssertTrue( ( 4 == Token_getOffset(pObj2)) );
        XCTAssertTrue( ( 3 == Token_getLineNo(pObj2)) );
        XCTAssertTrue( ( 2 == Token_getColNo(pObj2)) );
        XCTAssertTrue( (11 == Token_getClass(pObj2)) );
        XCTAssertTrue( (TOKEN_TYPE_INTEGER == Token_getType(pObj2)) );
        XCTAssertTrue( (64 == Token_getInteger(pObj2)) );

        obj_Release(pObj2);
        pObj2 = OBJ_NIL;
#endif

        obj_Release(pObj1);
        pObj1 = OBJ_NIL;
    }

    fprintf(stderr, "...%s completed.\n\n\n", pTestName);
    return 1;
}



int         test_Token_Test01(
    const
    char        *pTestName
)
{
    TOKEN_DATA    *pObj = OBJ_NIL;
    SRCLOC      loc = {1,2,0,3,4};

    fprintf(stderr, "Performing: %s\n", pTestName);
    pObj = Token_NewInteger(&loc, 11, 64);
    TINYTEST_FALSE( (OBJ_NIL == pObj) );
    if (pObj) {

        XCTAssertTrue( ( 1 == Token_getFileIndex(pObj)) );
        XCTAssertTrue( ( 4 == Token_getOffset(pObj)) );
        XCTAssertTrue( ( 3 == Token_getLineNo(pObj)) );
        XCTAssertTrue( ( 2 == Token_getColNo(pObj)) );
        XCTAssertTrue( (11 == Token_getClass(pObj)) );
        XCTAssertTrue( (64 == Token_getInteger(pObj)) );

        obj_Release(pObj);
        pObj = OBJ_NIL;
    }

    fprintf(stderr, "...%s completed.\n\n\n", pTestName);
    return 1;
}




TINYTEST_START_SUITE(test_Token);
    TINYTEST_ADD_TEST(test_Token_Test01,setUp,tearDown);
    TINYTEST_ADD_TEST(test_Token_Copy01,setUp,tearDown);
    TINYTEST_ADD_TEST(test_Token_OpenClose,setUp,tearDown);
TINYTEST_END_SUITE();

TINYTEST_MAIN_SINGLE_SUITE(test_Token);





